File: python/neuralgas.py
#import os,sys,inspect
import math
import nn
import random
import sys

class Node(nn.FeatureRepresentation):
    'doc'

    def __init__(self, features):
		nn.FeatureRepresentation.__init__(self,features);

    def computeSquaredFeatureDistance(self, sample):
		self.squaredFeaturesDistance=nn.FeatureRepresentation.computeSquaredFeatureDistance(self,sample)
		return self.squaredFeaturesDistance;

    def getSquaredFeaturesDistance(self):
		return self.squaredFeaturesDistance;

    def updateFeatures(self, e,  lamb, sample, k):
         h=math.exp(-k/lamb)
#         print `e*h`
         for i in range(0,len(sample)):
#             print i
             self.features[i]+=e*h*(sample[i]-self.features[i])
    def __str__(self):
          return `self.id`
    def __repr__(self):
          return self.__str__()
class Generator(nn.Generator):
    'doc'
        
    def generate(self,nodeCount,dimensionOfFeatures):
        nodes=range(nodeCount);
        for j in nodes:
            features=range(dimensionOfFeatures)
            for i in features:
                features[i]=random.random()
            nodes[j]=Node(features)
        return nodes;

class Network:
    'doc'
    
    def __init__(self,nodes):
        self.nodes=nodes;
        
    def init(self,tmax,estart,eend,lambdastart,lambdaend):
        self.t=0.0;
        self.tmax=tmax
        self.estart=estart
        self.eend=eend;
        self.lambdastart=lambdastart;
        self.lambdaend=lambdaend;

    def getT(self):
        return self.t;
        
    def learn(self, sample):
        featureDistMin=sys.float_info.max
        for node in self.nodes:
            dist=node.computeSquaredFeatureDistance(sample)
            if dist<featureDistMin:
                winner=node
                featureDistMin=dist
#        print self.nodes[0].getSquaredFeaturesDistance()
        self.nodes.sort(key=lambda x: x.squaredFeaturesDistance, reverse=False)
#        print self.nodes        
        timeq=self.t/self.tmax;
        e=self.estart*(pow(self.eend/self.estart, timeq));
        lambd=self.lambdastart*(pow(self.lambdaend/self.lambdastart, timeq));
        i=0;
        for node in self.nodes:
            node.updateFeatures(e, lambd,sample,i);
#            print i
            i=i+1
#       print '----'+`len(self.nodes)`
        self.t+=1;
#        print '--'+`e`+' '+`(pow(self.eend/self.estart, timeq))`+' '+`self.eend/self.estart`+' '+`timeq`+' '+`self.t`+' '+`self.tmax`

    def getNodes(self):
        return self.nodes

File: c++/nn_multi_lang/neuralgas.cpp
#include "neuralgas.hpp"
#include <math.h>
#include <limits>
#include <algorithm>
#include <iostream>

using namespace nn;
using namespace std;

namespace neuralgas
{
    Node::Node(std::vector<double> features):FeatureRepresentation(features)
    {
    }
    Node::~Node()
    {
    }
    double Node::getSquaredFeaturesDistance()
    {
        return squaredFeaturesDistance;
    }
 	double Node::computeSquaredFeatureDistance(std::vector<double> sample)
	{
		squaredFeaturesDistance=FeatureRepresentation::computeSquaredFeatureDistance(sample);
		return squaredFeaturesDistance;
	}
   void Node::updateFeatures(double e, double lambda,std::vector<double> sample,double k)
    {
        double h=exp(-k/lambda);
//        cout<<(e*h)<<endl;
        for (unsigned int i=0;i<sample.size();++i)
        {
            features[i]+=e*h*(sample[i]-features[i]);
        }
    }
 	std::vector<Node> Generator::generate(unsigned int nodeCount, unsigned int dimensionOfFeatures)
	{
		std::vector<Node> nodes;
		for(unsigned int j=0;j<nodeCount;++j)
		{
			std::vector<double> features=nn::Generator::generateSample(dimensionOfFeatures);
			nodes.push_back(Node(features));
		}
		return nodes;
	}
	Network::Network(std::vector<Node> n):nodes(n)
	{
	}
	void Network::init(double _tmax, double _estart, double _eend, double _lambdastart, double _lambdaend)
	{
		t=0;
		this->tmax=_tmax;
		this->estart=_estart;
		this->eend=_eend;
		this->lambdastart=_lambdastart;
		this->lambdaend=_lambdaend;
	}
	double Network::getT()
	{
		return t;
	}
    bool sortImpl(Node left, Node right)
    {
        return left.getSquaredFeaturesDistance()<right.getSquaredFeaturesDistance();
    }
	void Network::learn(std::vector<double> sample)
	{
		double featureDistMin=std::numeric_limits<double>::max();
		cout<<featureDistMin<<endl;
		Node &winner=nodes[0];

		for (unsigned int i=0;i<nodes.size();++i)
		{
			double dist=nodes[i].computeSquaredFeatureDistance(sample);
//			cout << i<< " "<<dist<< " "<<(dist<featureDistMin)<<endl;
			if(dist<featureDistMin)
			{
				winner=nodes[i];
				featureDistMin=dist;
			}
		}
//cout<<winner.getId()<<endl;
cout<<featureDistMin<<" "<<winner.getId()<<" "<<nodes[0].getId()<< endl;

		std::sort(nodes.begin(),nodes.end(),sortImpl);
cout<<winner.getId()<<" "<<nodes[0].getId() << endl;
		double timeq=t/tmax;
		double e=estart*(pow(eend/estart, timeq));
		double lambda=lambdastart*(pow(lambdaend/lambdastart, timeq));
//		cout<<t<<" "<<timeq<<" "<<e<<" "<<lambda<<endl;
		for (unsigned int i=0;i<nodes.size();++i)
		{
			nodes[i].updateFeatures(e, lambda,sample,i);
		}
		t+=1;
	}
	std::vector<Node> Network::getNodes()
	{
		return nodes;
	}
}
